snetd/cmd/serve.go

```diff
@@ -81,34 +96,71 @@ func newDaemon() (daemon, error) {
 
 func (d daemon) start() {
 	d.blockProc.StartLoop()
+	var tlsConfig *tls.Config
+
+	if d.autoSSLDomain != "" {
+		certMgr := autocert.Manager{
+			Prompt:     autocert.AcceptTOS,
+			HostPolicy: autocert.HostWhitelist(d.autoSSLDomain),
+			Cache:      autocert.DirCache(config.GetString(config.AutoSSLCacheDirKey)),
+		}
+
+		// This is the HTTP server that handles ACME challenge/response
+		acmeSrv := http.Server{
+			Handler: certMgr.HTTPHandler(nil),
+		}
+		go acmeSrv.Serve(d.acmeListener)
+
+		tlsConfig = &tls.Config{
+			GetCertificate: func(c *tls.ClientHelloInfo) (*tls.Certificate, error) {
+				crt, err := certMgr.GetCertificate(c)
+				if err != nil {
+					log.WithError(err).Error("unable to fetch certificate")
+				}
+				return crt, err
+			},
+		}
+	}
 
 	if config.GetString(config.DaemonTypeKey) == "grpc" {
-		mux := cmux.New(d.lis)
-		grpcL := mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldPrefixSendSettings("content-type",
-			"application/grpc"))
+		d.grpcServer = grpc.NewServer(
+			grpc.UnknownServiceHandler(handler.GetGrpcHandler()),
+			grpc.StreamInterceptor(d.blockProc.GrpcStreamInterceptor()),
+		)
+
+		lis := d.lis
+		if tlsConfig != nil {
+			log.Debug("enabling SSL support")
+			tlsc := tlsConfig.Clone()
+			// See: https://gist.github.com/soheilhy/bb272c000f1987f17063
+			tlsc.NextProtos = []string{http2.NextProtoTLS, "h2-14", "http/1.1"}
+
+			lis = tls.NewListener(lis, tlsc)
+		}
+		mux := cmux.New(lis)
+		grpcL := mux.Match(cmux.HTTP2HeaderField("content-type", "application/grpc"))
```

Comment:
We have this on prefix matching as content-type may be one of `application/grpc`, `application/grpc+proto`, `application/grpc+json` for our grpc clients

And we use MatchWithWriters/*SendSettings I think just for compatibility with Java clients... https://github.com/soheilhy/cmux#limitations 

I should have commented on both of these decisions specifically

Reply:
Makes sense. I changed a bunch of this stuff while testing to try different permutations of protocols and didn't change it back to how it was originally.

Updated to revert changes and add comment as to why we're doing the prefix/settings stuff.